libfwupd/fwupd-common.c

```diff
@@ -133,3 +135,115 @@ fwupd_checksum_get_best (GPtrArray *checksums)
 	}
 	return NULL;
 }
+
+static gchar *
+fwupd_build_user_agent_os_release (void)
+{
+	const gchar *keys[] = { "NAME=", "VERSION_ID=", "VARIANT=", NULL };
+	const gchar *values[] = { NULL, NULL, NULL, NULL };
+	g_autofree gchar *os_release = NULL;
+	g_auto(GStrv) lines = NULL;
+	g_autoptr(GPtrArray) ids_os = g_ptr_array_new ();
+
+	/* get raw data then parse each line */
+	if (!g_file_get_contents ("/etc/os-release", &os_release, NULL, NULL))
+		return NULL;
+	lines = g_strsplit (os_release, "\n", -1);
+	for (guint i = 0; lines[i] != NULL; i++) {
+		for (guint j = 0; keys[j] != NULL; j++) {
+			if (g_str_has_prefix (lines[i], keys[j])) {
+				values[j] = lines[i] + strlen (keys[j]);
+				break;
+			}
+		}
+	}
+
+	/* create an array of the keys that exist */
+	for (guint j = 0; values[j] != NULL; j++) {
+		if (values[j] != NULL)
+			g_ptr_array_add (ids_os, (gpointer) values[j]);
+	}
+	if (ids_os->len == 0)
+		return NULL;
+	g_ptr_array_add (ids_os, NULL);
+	return g_strjoinv (" ", (gchar **) ids_os->pdata);
+}
+
+static gchar *
+fwupd_build_user_agent_system (void)
+{
+	struct utsname name_tmp = { 0 };
+	g_autofree gchar *locale = NULL;
+	g_autofree gchar *os_release = NULL;
+	g_autoptr(GPtrArray) ids = g_ptr_array_new_with_free_func (g_free);
+
+	/* system, architecture and kernel, e.g. "Linux i686 4.14.5" */
+	if (uname (&name_tmp) >= 0) {
+		g_autofree gchar *kernel_version = g_strdup (name_tmp.release);
+		g_strdelimit (kernel_version, "-", '\0');
```

Comment:
I don't think it's a good idea to strip out distro kernel build number. Sometimes large patches are backported and you wouldn't be able to cross reference distro change log to tell if they are there.

Reply:
Makes sense I guess, fixed.